### PR TITLE
quiche: enable downstream HTTP3 in quic_protocol_integration_test

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: Enable protocol integration test for HTTP3
Additional Description: some tests are disabled; Added quic client connection factory to HttpIntegrationTest to create quic client connection; Move the config setup logic from quic_http_integration_test to ConfigHelper and used it in the base class HttpIntegrationTest.

Add fixes in server and client streams to make some tests pass:
Check whether `:authority` is header has valid host name.
Set `:protocol` and `:path` field in request headers if the `:method` is CONNECT.
Quic trailers concatenates header values with the same key with '\0', split them into multiple headers.

Risk Level: low

Part of #12930 #2557

### CODE REVIEW

File: source/extensions/quic_listeners/quiche/envoy_quic_utils.cc
Comment (alyssawilk) [inline source/extensions/quic_listeners/quiche/envoy_quic_utils.cc:122]: Should this be loopback?
I'd think
local_addr = Network::Utility::getLocalAddress(host_address->ip()->version());
would make more sense though you may have to do some v6 workarounds for tests as I did in #15027

File: test/integration/protocol_integration_test.cc
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:198]: Fine to leave this disabled for now but I think we can adapt the test filter to change a request with no body (headers, fin) to do (headers, fin, body).  I've adapted a few filters in my test PR so might as well wait for that to land.

File: test/integration/protocol_integration_test.cc
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:445]: some of the header tests should be fixed by 15381 but others need configurable header size to pass. We'll have to figure that one out at some point :-)

File: test/integration/protocol_integration_test.cc
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:1115]: TODO to add a stat for H3?

File: test/integration/protocol_integration_test.cc
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:1362]: huh, this should probably fail.  Your plan is to address all the ones without TODOs so there's no need for an explicit TODO here right?

File: test/integration/http_integration.cc
Comment (alyssawilk) [inline test/integration/http_integration.cc:315]: would it be possible to have this either take options or share code from the SSL function, so if we need to change TLS defaults we only have to do it in one place?

File: test/integration/http_integration.cc
Comment (alyssawilk) [inline test/integration/http_integration.cc:1053]: preference for quic logic in the normal pause filter
I used the connection ... protocol() as a detect option for mine, may work for you?

**Review by alyssawilk — COMMENTED** [review]
Awesome!  So good to get more coverage for HTTP/3!

File: test/integration/http_integration.cc
Comment (danzh2010) [reply test/integration/http_integration.cc:1053]: pause filter library is depended by http_integration_lib, we want them to be able to compile out QUICHE right?

Comment (alyssawilk) [reply test/integration/http_integration.cc:1053]: ah yeah, so merging it in will have to wait until the compile macros land, good point.
do we really need all the custom down-casting rather than just forcing callbacks?  I had missed the degree of quic-specific logic here.

File: test/integration/protocol_integration_test.cc
Comment (alyssawilk) [inline test/integration/protocol_integration_test.cc:1981]: you probably need to tweak times_added.
AFIK bytes added in trailers only come through for H2/H3 on both sides so times_added should be 2 if trailers work for both upstream and downstrema.

File: test/integration/ssl_utility.cc
Comment (alyssawilk) [inline test/integration/ssl_utility.cc:60]: ?

**Review by alyssawilk — COMMENTED** [review]
This is so awesome!  we're really getting close to everything working for HTTP/3
please do checkout the TSAN ci failure as you address comments.

/wait

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (danzh2010) [reply test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:124]: Nope, SAN is set to the default value if not specified in SslUtility. TCP TLS tests so far didn't use any other value.

File: source/extensions/quic_listeners/quiche/active_quic_listener.cc
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/active_quic_listener.cc:45]: I couldn't find the crash stack now. But it is a QUIC_BUG packet serialization failure for packet with ACK and CONNECTION_CLOSE frame. It is reported as internal bug: 173790064 and fixed by quic team.

File: test/integration/http_integration.cc
Comment (danzh2010) [reply test/integration/http_integration.cc]: It only takes reuse_port_ and can be made into ConfigHelper.

File: test/integration/http_protocol_integration.h
Comment (danzh2010) [reply test/integration/http_protocol_integration.h]: Added an argument downstream_use_quic to ConfigHelper::httpProxyConfig() and branches inside of it.

File: test/integration/protocol_integration_test.cc
Comment (danzh2010) [reply test/integration/protocol_integration_test.cc:1981]: The failed tests uses H1 for downstream, so I would expect times_added should be 1, but the test fails with:
[ RUN      ] UpstreamProtocols/DownstreamProtocolIntegrationTest.TestEncodeHeadersReturnsStopAllWatermark/IPv4_HttpDownstream_Http3Upstream
test/integration/protocol_integration_test.cc:1984: Failure
Expected equality of these values:
  count_ * size_ + added_decoded_data_size_ * times_added
    Which is: 70001
  response->body().size()
    Which is: 70002

So I suspect Http3 somehow added the twice.

File: source/extensions/quic_listeners/quiche/active_quic_listener.cc
Comment (DavidSchinazi) [reply source/extensions/quic_listeners/quiche/active_quic_listener.cc:45]: Thanks for the details Dan. This is a different crash from the one I'm investigating.

**Comment by danzh2010** [comment]
The TSAN keeps failing at integration_admin_test: https://dev.azure.com/cncf/envoy/_build/results?buildId=69583&view=logs&j=d1f76054-8f79-554b-6f4a-11d6a63b8e00&t=266e17e3-d213-54b5-deef-0dcee01da137&l=15243. Seems unrelated?

File: source/extensions/quic_listeners/quiche/envoy_quic_utils.h
Comment (jmarantz) [inline source/extensions/quic_listeners/quiche/envoy_quic_utils.h:64]: I was surprised to see splitting on `\0` and so I went looking around and found the definition in spdy/core/spdy_header_block.h which confirms it, but not all the time:
```
  // If a header with the key is already present, then append the value to the
  // existing header value, NUL ("\0") separated unless the key is cookie, in
  // which case the separator is "; ".
  // If there is no such key, a new header with the key and value is added.
  void AppendValueOrAddHeader(const absl::string_view key,
                              const absl::string_view value);
```
I've never worked with that library before but I wonder if it supplies some sort of split API we should be using rather than calling StrSplit directly.

File: test/config/utility.cc
Comment (jmarantz) [inline test/config/utility.cc:1065]: should we be checking that filter_chains is non-empty?

File: test/integration/utility.cc
Comment (jmarantz) [inline test/integration/utility.cc:90]: RE questions asked over DM about integration_admin_test, skimming over your PR it looks like this code might be interesting for some breakpoints.

I don't have a clear picture of why it's needed, so maybe some class comments would help.

File: test/integration/utility.cc
Comment (jmarantz) [inline test/integration/utility.cc:143]: this function has gotten a lot longer and harder to follow due to the handling of h3 in it. Can we think of ways to refactor this so that the function structure remains more like it used to be?

E.g. one option would be to just add an early dispatch:
```
  if (h3 mode) {
    return makeSingleH3Request(...);
  }
```
and then factor out the common blocks from the h3 and non-h3 impls into some helpers.

WDYT?

File: test/integration/utility.cc
Comment (alyssawilk) [inline test/integration/utility.cc:102]: I'm unclear on why we need this class.
Does TLS not test connection failures?
Can you snag me on IM and point me at where we're hanging and perhaps I can suggest an alternate fix?

File: source/extensions/quic_listeners/quiche/envoy_quic_utils.h
Comment (jmarantz) [reply source/extensions/quic_listeners/quiche/envoy_quic_utils.h:64]: I was mostly just thinking that the \0 delim looks like an implementation detail of spdy_header_block.h and the detail should ideally be abstracted away at that level.

And yes I was also wondering specifically about what the behavior would be with cookies.

Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/envoy_quic_utils.h:64]: Unfortunately, SpdyHeaderBlock doesn't have such API to split the `\0` delimited header values. Even quic qpack code has to write its own logic to handle this special case.

File: test/integration/utility.cc
Comment (danzh2010) [reply test/integration/utility.cc:174]: Thanks for suggesting the refactoring! Surprisingly, this fix the TSAN issue I asked earlier. I found that createQuicUpstreamTransportSocketFactory() which was also called under h2 and h1 earlier somehow breaks the stats histogram Expectation. But I still don't understand how could a client util function lead to server stats change. This function calls a registered factory to create a quic upstream transport socket factory for the test client.

File: test/config/utility.cc
Comment (danzh2010) [reply test/config/utility.cc:1065]: This is just a test util function whose the input is the config template defined in the same file. But I added ASSERT to help future debugging.

File: test/integration/utility.cc
Comment (jmarantz) [reply test/integration/utility.cc:174]: Yeah the reason is that went looking at the code in its previous state, it was very heard to discern whether the behavior was the same in H2 with/without your change. I suspect it wasn't, and that was what was affecting the integration test tsan.

Comment (jmarantz) [reply test/integration/utility.cc]: I'd guess this is inducing the client to change the way it communicates with the server in a stat-observable way, but I'd have to dig in with the debugger to learn more.

File: source/extensions/quic_listeners/quiche/envoy_quic_utils.h
Comment (jmarantz) [reply source/extensions/quic_listeners/quiche/envoy_quic_utils.h:64]: Some comments here might help at least clear up the mysticism, if it's really impossible to create an abstraction boundary around this.

File: test/integration/utility.cc
Comment (alyssawilk) [reply test/integration/utility.cc:108]: can we get a TODO to clean this up?  I think I'd prefer we rework the test framework to handle this failure mode than have this test class. I suspect there's a cleaner way but I don't want to block the PR on it.

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo the two requested nits.
@mattklein123 up for a quick sweep as well?

File: test/integration/utility.cc
Comment (danzh2010) [reply test/integration/utility.cc:110]: TODO added.

File: test/integration/utility.cc
Comment (danzh2010) [reply test/integration/utility.cc]: I reproduced the same failure by adding a line to print trace in a clean branch https://github.com/envoyproxy/envoy/compare/main...danzh2010:admin_test?expand=1. Is there some existing race in that test?

Comment (jmarantz) [reply test/integration/utility.cc]: Interesting. Would a sleep there also cause the test to fail? How about printing a bunch of text to stderr?

That trace macro doesn't look like it ought to wake up anything in stats to change behavior, so I suspect a sleep or simple I/O might do the same thing.

**Review by alyssawilk — APPROVED** [review]

File: source/extensions/quic_listeners/quiche/active_quic_listener.cc
Comment (mattklein123) [inline source/extensions/quic_listeners/quiche/active_quic_listener.cc:45]: Should this be a TODO and/or link to a follow up issue? Do we plan on removing this later?

File: source/extensions/quic_listeners/quiche/client_connection_factory_impl.cc
Comment (mattklein123) [inline source/extensions/quic_listeners/quiche/client_connection_factory_impl.cc:32]: TODO/link?

File: source/extensions/quic_listeners/quiche/envoy_quic_utils.h
Comment (mattklein123) [reply source/extensions/quic_listeners/quiche/envoy_quic_utils.h:64]: +1 for more comments

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Exciting to see so much active progress on H3! Just some small comments on source.

/wait

File: source/extensions/quic_listeners/quiche/active_quic_listener.cc
Comment (alyssawilk) [reply source/extensions/quic_listeners/quiche/active_quic_listener.cc:45]: quiche flags are like our runtime guards only they start false - it'll move to true by default and get removed over some number of import cycles.

**Review by alyssawilk — APPROVED** [review]

File: source/extensions/quic_listeners/quiche/envoy_quic_utils.h
Comment (danzh2010) [reply source/extensions/quic_listeners/quiche/envoy_quic_utils.h:65]: done

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
